### ISSUE TITLE
Quantized model has different output between pytorch and onnx

### ISSUE THREAD

**Issue by lyrstc** [NVIDIA/TensorRT#2354]
## Description

We quantize the bert model using pytorch-quantization as the https://github.com/NVIDIA/TensorRT/blob/main/tools/pytorch-quantization/examples/calibrate_quant_resnet50.ipynb.

Then, we export the quantized model to onnx using torch.onnx.

However, the output has a obvious difference betwwen onnx and pytorch output.

For bert-base model, The absolute difference of output logit is over 0.01, and the difference of hidden state could reach 0.1. The code is as below, and the output of each layer is represented. The error seems small for the first several layers, and suddenly become great.

There is warning when export quantized model to onnx format, but I don't think that's the key reason. the warning message is as follow:

W0926 10:45:13.264863 140686472021760 tensor_quantizer.py:280] Use Pytorch's native experimental fake quantization. /mnt3/baize.lyr/anaconda3/envs/rt_b/lib/python3.8/site-packages/pytorch_quantization/nn/modules/tensor_quantizer.py:283: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs! if amax.numel() == 1: /mnt3/baize.lyr/anaconda3/envs/rt_b/lib/python3.8/site-packages/pytorch_quantization/nn/modules/tensor_quantizer.py:285: TracerWarning: Converting a tensor to a Python number might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs! inputs, amax.item() / bound, 0, /mnt3/baize.lyr/anaconda3/envs/rt_b/lib/python3.8/site-packages/pytorch_quantization/nn/modules/tensor_quantizer.py:291: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs! quant_dim = list(amax.shape).index(list(amax_sequeeze.shape)[0])

Another discovery is that when the input attention_mask is a randomly generated 0 or 1 tensor, the error will be greater.

## Environment

TensorRT：8.4.1.5   
gpu driver：515.48    
cuda：11.7，  
pytorch:1.12     
onnx:1.11  
python:3.8  
pytorch-quantization:2.1.2  
onnxruntime 1.11.1  

## Relevant Files

[issue_eng.html.zip](https://github.com/NVIDIA/TensorRT/files/9643774/issue_eng.html.zip)

## Steps To Reproduce

the error is shown in detail in the above jupyter notebook

**Comment by zerollzeng** [comment]
@ttyio ^ ^

**Comment by ttyio** [comment]
Hello @lyrstc , besides numeric comparison, have you measured the benchmark score. e.g, the F1 score on SQuAD dataset? Thanks!

**Comment by lyrstc** [comment]
> Hello @lyrstc , besides numeric comparison, have you measured the benchmark score. e.g, the F1 score on SQuAD dataset? Thanks!

I have measured the bert-base(downloaded from huggingface "textattack/bert-base-uncased-MRPC") on the MRPC dataset, the accuracy of the quantized pytorch model(calibration with max calibrator) is 0.84, however, after export the model to onnx format, the accuracy drop to 0.32. 

The error is much larger in the real data. It might because that the 'attention_mask' and 'token_type_ids' are set to be 1 in simulation data but tensors include (0,1) in real data.

**Comment by ttyio** [comment]
@lyrstc , I go through the notebook, here we calibration and QAT using a random data. Have you experiment using a real dataset? thanks!

**Comment by lyrstc** [comment]
> @lyrstc , I go through the notebook, here we calibration and QAT using a random data. Have you experiment using a real dataset? thanks!

The code of experiment on the MRPC dataset is represent in the following notebook. As i mentioned before, the error seems much larger. And i think the larger difference may be related to the input of 'attention_mask' 'token_type_ids'.
[issue_mrpc.html.zip](https://github.com/NVIDIA/TensorRT/files/9652116/issue_mrpc.html.zip)

**Comment by ttyio** [comment]
> > @lyrstc , I go through the notebook, here we calibration and QAT using a random data. Have you experiment using a real dataset? thanks!
> 
> The code of experiment on the MRPC dataset is represent in the following notebook. As i mentioned before, the error seems much larger. And i think the larger difference may be related to the input of 'attention_mask' 'token_type_ids'. [issue_mrpc.html.zip](https://github.com/NVIDIA/TensorRT/files/9652116/issue_mrpc.html.zip)

Thanks! It seems there is bug in either the torch to onnx export, or the onnxruntime. 
Have you tried TensorRT inference on this onnx model? Is the result close to pytorch or to onnxruntime?

**Comment by lyrstc** [comment]
> 



> > > @lyrstc , I go through the notebook, here we calibration and QAT using a random data. Have you experiment using a real dataset? thanks!
> > 
> > 
> > The code of experiment on the MRPC dataset is represent in the following notebook. As i mentioned before, the error seems much larger. And i think the larger difference may be related to the input of 'attention_mask' 'token_type_ids'. [issue_mrpc.html.zip](https://github.com/NVIDIA/TensorRT/files/9652116/issue_mrpc.html.zip)
> 
> Thanks! It seems there is bug in either the torch to onnx export, or the onnxruntime. Have you tried TensorRT inference on this onnx model? Is the result close to pytorch or to onnxruntime?

The code is as follows.  I added the inference result of Tensorrt ( 'transformer_deploy' code is used, but it will not affect the result). From the results, the outputs of Tensorrt are close to onnx' outputs. However, as the layer gets deeper, the error will also increase. The error of the final logit is within 0.1. 

I wonder if this error accumulation is normal?
[issue_trtresult.html.zip](https://github.com/NVIDIA/TensorRT/files/9652732/issue_trtresult.html.zip)

**Comment by ttyio** [comment]
@lyrstc , I checked your `issue_eng.html` output again, the diff is:

        ([-1.77020359,  2.08652067])
        [[-1.798069   2.0711205]]

and since your experiment show that onnxruntime is more close to trt result. I am thinking of this is as float arithmetic error.
Have you also checked the onnx model without QAT? Thanks!

**Comment by lyrstc** [comment]
> @lyrstc , I checked your `issue_eng.html` output again, the diff is:
> 
> ```
>     ([-1.77020359,  2.08652067])
>     [[-1.798069   2.0711205]]
> ```
> 
> and since your experiment show that onnxruntime is more close to trt result. I am thinking of this is as float arithmetic error. Have you also checked the onnx model without QAT? Thanks!

thank you for your reply, i don't use QAT in the previous code. I think i just added the QDQ node using pytorch_quantization and applied calibration. 

Besides, I want to confirm with you whether a pytorch model with QDQ nodes will have numeric error in any case when it is converted to onnx and tensorrt? Is the accuracy loss caused by conversion unavoidable, and what is the order of magnitude of numerical error in general?

**Comment by ttyio** [comment]
closing due to no activity for more than 3 weeks, please reopen if you still have question, thanks!

**Comment by lucy3589** [comment]
hi, I had a similar problem
qat_yolov8n-seg model, val .onnx mAP is slower than val .pt

qat：
from pytorch_quantization import nn as quant_nn
from pytorch_quantization import quant_modules
quant_nn.TensorQuantizer.use_fb_fake_quant = True
quant_modules.initialize()

train_model, get yolov8n-seg-qat.pt

#pt2onnx
from ultralytics import YOLO
model = YOLO("yolov8n-seg-qat.pt")val_model
success = model.export(format="onnx", simplify=True, imgsz=640,opset=13,device=0)
assert success

#val_model
from ultralytics import YOLO
#model = YOLO('yolov8n-seg-qat.pt')
model = YOLO('yolov8n-seg-qat.onnx')
model.task = 'segment'
metrics = model.val()
metrics.box.map
metrics.box.map50
metrics.box.map75
metrics.box.maps
metrics.seg.map
metrics.seg.map50
metrics.seg.map75
metrics.seg.maps
print(metrics)

but onnx_metrics is poor than pt_metrics

.pt accuary is higher than .onnx 3%

**Comment by ZHAOZHIHAO** [comment]
@lucy3589  See if this is related https://discuss.pytorch.org/t/how-to-export-a-correct-quantized-model-to-onnx-format/186322
